src/Adapter/Product/ProductProvider.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product;
+
+use PrestaShop\PrestaShop\Adapter\AbstractObjectModelProvider;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use Product;
+
+/**
+ * Provides existing Product
+ */
+class ProductProvider extends AbstractObjectModelProvider
+{
+    /**
+     * @param ProductId $productId
+     *
+     * @return Product
+     *
+     * @throws CoreException
+     */
+    public function get(ProductId $productId): Product
+    {
+        /** @var Product $product */
+        $product = $this->getObjectModel(
```

Comment:
Oh I just thought of something, it doesn't need to be managed here but the problem will probably arise at some point

The Product object has an additional `full` parameter in its constructor, and we will definitely need it sometimes (I don't know if it should be optional or always full just to be sure) To my knowledge it's the only ObjectModel with such additional constructor field but me might have to handle a special case for it at some point

Maybe allow optional additional parameters in the `AbstractObjectModelProvider` later or something.. we'll see ^^

Reply:
btw it also accepts `$id_lang = null, $id_shop = null, Context $context = null`, but we haven't used any of this so far :smile:

Reply:
No indeed, I think `$id_lang` must remain null, when it's set the localized fields are transformed into strings automatically We don't want this for edition
The shop might be useful though when we will deal with mutlishop